Aidan-improvements-layerbbox

aidanwobrien opened the pull request:
I added a way for a user to choose whether they want to query the service using the bounding box of the canvas (existing) or the bounding box of the active layer (new)

At the top of the layer multi select dialog box you will see "Use selected layer for bbox" or "Use canvas for bbox"

By default, using the canvas for bbox is selected.

This meant adding a new function to extract the bbox from the active layer and running that function if the user selects "Use selected layer for bbox". 

Also added some error handling in that new function for if no active layer is selected.

aidanwobrien committed:
WIP adding option to use a layer bbox for query

aidanwobrien committed:
Error handling is no layer is selected

aidanwobrien committed:
clarifying comments

aidanwobrien commented:
Some potential issues i see here 

1) the users active layer is their basemap - that will return a massive bounding box and jam up the API
2) the users active layer is a point or line... that will probably mess up the bounding box too and might throw an error (yet to test this)

I think I can easily address this by checking that the active layer is 1) a vector and 2) a polygon or multipolygon. So this functionality would be limited to using vector polygon layers.

aidanwobrien committed:
added error handling
to query a selected layers bbox, it must be polygon

aidanwobrien commented:
I added a check to make sure the active layer is indeed a polygon

aidanwobrien committed:
Update README.md

added a brief "how-to" on using the main load_lyrs_from_geohub_services script

aidanwobrien committed:
Added a clipping function that clips the results
(queried from the active layers bbox)
to each feature in the active layerbbox

Also edited the Building As Symbol layer
in the lio_list and styles files
to match the Ontario Geohub
(Building as Symbol)
So that layer is styled correctly now

aidanwobrien committed:
Added Ontario Trail Network
Segment (line) and Access Point (point) layers
to lio_list.py and to styles.py

aidanwobrien committed:
updating save_and_log_sel file to address conflict

frasemcl reviewed (changes requested):
Thanks for all this, very helpful! Will your logic work if the following block also accepts MultiPolygon? I ran into an error on my first run due to file format.
```
            # Active layer must be a polygon
            elif QgsWkbTypes.displayString(active_layer.wkbType()) != "Polygon":
                print("The selected layer needs to be a polygon!")
                iface.messageBar().pushMessage("Error", "Selected layer is not a polygon!", level=Qgis.Critical)
                raise ValueError("Selected layer is not a polygon!")
```

aidanwobrien commented:
Good catch - I dont think it will work with MultiPolygon at the moment. Let me take a look.

Was the polygon you were using a MultiPolygon? @frasemcl

aidanwobrien committed:
update to handle multi-polygons

aidanwobrien committed:
simplify multipolygon error handling

aidanwobrien commented:
I added this simple logic to accept multi-polygons.

```
            elif QgsWkbTypes.displayString(active_layer.wkbType()) == "MultiPolygon":
                pass

```

aidanwobrien committed:
making the error handling more logical

aidanwobrien committed:
simplify the error handling even more

aidanwobrien commented:
I simplified things to be easier to read. The error handling looks like this now

```
            if not active_layer:
                print("No layer selected!")
                iface.messageBar().pushMessage("Error", "No layer selected!", level=Qgis.Critical)
                raise ValueError("No layer selected!")
                # return

            # If the active layer is a raster, throw an error
            elif active_layer.type() == QgsMapLayer.RasterLayer:
                print("The selected layer is a raster!")
                iface.messageBar().pushMessage("Error", "Selected layer is a raster!", level=Qgis.Critical)
                raise ValueError("Selected layer is a raster!")

            # If the active layer is a multi-polygon or polygon, keep going!
            elif QgsWkbTypes.displayString(active_layer.wkbType()) in ["MultiPolygon", "Polygon"]:
                pass

            # If the active layer is not a Polygon or MultiPolygon, throw an error
            else:
                print("The selected layer needs to be a polygon!")
                iface.messageBar().pushMessage("Error", "Selected layer is not a polygon!", level=Qgis.Critical)
                raise ValueError("Selected layer is not a polygon!")
```

frasemcl reviewed (commented):
To answer your previous question, I did try to use the new feature with a multipolygon (without realizing) the first time I tested this. I like the solution, BUT, can you confirm that the load from polygon feature is working for you? It's going very slowly for me when I try a polygon or a multipolygon. It worked with viewport extent but that was also slower than usual today.

frasemcl commented:
Seems like the bug causing the script to be very slow for me is that when I click 'Run', my project CRS switches to WGS84. I typically run the script in UTM Zone 17N or NAD83 and don't remember this ever happening. I do remember some transforms being really slow (or not working at all) in my initial development of this.

aidanwobrien committed:
removed setting project to epsg:4326
use canvas extent as bbox instead of active_layer
add timer to monitor script execution time

aidanwobrien commented:
Sorry this is ending up being so confusing. In short, there are some complexities I'm not considering.

I did initially add a line to set the project crs to ESPG:4326... that is not needed. I need to adjust my logic for how the layer_bbox is determined.

I need to adjust my layer_bbox_for_service function which currently queries only the bbox of each feature using the active_layer.exent(). But I think what is slowing things down so much is the polygon layers CRS can be different than the project crs. 

Things would be more simple if the users canvas always had the extent of the polygons layer... we could just re-use the canvas_bbox_for_service function and clip the results to the polygons.... but I doubt that will always be the case.

So instead I think I need to transform the active_layer to match the project CRS... which still might slow things down some.

I'm back to the drawing board, so feel free to reject this PR and I'll keep working on it.

aidanwobrien committed:
transformation bug fix

aidanwobrien committed:
overlay layer bug fix

aidanwobrien committed:
clipping function bug fix

aidanwobrien committed:
making the feature counts faster

aidanwobrien committed:
remove pyqgis_group if the user cancels the script

aidanwobrien committed:
fix invalid geometries before clipping

frasemcl commented:
Closed for now, I see you're still working on this logic so I'll know you're ready to show me when you open it again.
